paddle/fluid/operators/run_program_op.h

```diff
@@ -149,14 +153,44 @@ static void ShareVarsFromScope(const std::vector<Variable *> &vars,
   }
 }
 
-static void AppendSkipDeletionVars(
-    std::vector<std::string> *all_vars,
-    const std::vector<std::string> &append_vars) {
+static void AppendSkipDeletionVars(const std::vector<std::string> &append_vars,
+                                   std::vector<std::string> *all_vars) {
   for (auto &var : append_vars) {
     all_vars->emplace_back(var);
   }
 }
 
+static void AppendSafeEagerDeletionSkipVars(
+    const framework::ProgramDesc &program,
+    std::vector<std::string> *skip_vars) {
+  const framework::BlockDesc &block = program.Block(0);
+  const std::vector<framework::OpDesc *> &all_ops = block.AllOps();
+
+  std::unordered_set<std::string> grad_op_output;
+  std::unordered_set<std::string> grad_op_input;
+  for (const framework::OpDesc *op : all_ops) {
+    if (!boost::algorithm::ends_with(op->Type(), "_grad")) {
+      continue;
+    }
```

Comment:
I have  a question:  here we use  `_grad` to determine whether a op is `grad_op`, but some op's grad_op is not with `_grad` suffix, such `arrary_write`'s grad_op is `array_read`.

Reply:
maybe also can use `op_role`?

Reply:
Thank you for great suggestion. Done